README.md
This app is only able to function if you have the Radiation Detector with you.
You may be able to register an account and view the radiation readings recorded by other users, but you will not be able to record or send data to the database.

The app was coded using Flutter for cross platform deployment (Appstore and Google Play Store).
It is using the DART framewrok and has been intergrated with Firebase for it's data storage.
This app is a collaboration with my teammate, otakenz(Github profile).

This project was conducted under the internship at Singapore Nuclear Research and Safety Initiative,(SNRSI). This code was delivered to them, and they have the right to alter and further develop future versions of this app.


CHANGELOG.md
## 1.2.0 (August 03, 2020)

- Merge pull request #33 from Nntsyeo/RAD-0022
- [RAD-0022] changed the secondary color
- [RAD-0022] Black
- [RAD-0022] Fix "u" and map
- [RAD-0022] Fixed long init on graph
- Removed dark theme final
- Remove Dark Theme
- Removed Dark Theme and Added Color Theme
- Fixed overflow issues
- Added Building Level Input
- update Battery Status persecond
- Completed All indicators and alert, Current issue, Page takes long tiem for performance setup
- Added Location and Weather in data display page, Successfully sends to Firebase firestore too
- Merge pull request #32 from Nntsyeo/RAD-0035
- Completed Verification and Navigation between Sign up and Sign In and HomePage
- Login logic reworked, added reverify email and reset password
- worked on logic flow
- added Verifier and email can always be sent. Image changes to EIC logo. Currently verification takes forever
- working on verification token"
- email verifier added
- Merge pull request #31 from Nntsyeo/RAD-0016
- Disabled the underline function for the content
- Added the url link to the website and the email
- Updated the information on educational page
- Added scrollable feature on education page
- Added Titles for Education Page
- Merge pull request #30 from Nntsyeo/RAD-0048
- [RAD-0048] Accomodate new data entry
- Merge pull request #28 from Nntsyeo/RAD-0045
- [RAD-0045] Finalized responsive ui
- Initial responsive ui
- [RAD-0045] testing
- [RAD-0045] Initial implementation of responsive UI
- Merge branch 'release-1.1.0+7' into develop
- Bumped version number to 0.2.0

## 0.2.0 (July 30, 2020)

## 0.1.0 (July 30, 2020)

- Merge pull request #27 from Nntsyeo/RAD-0006
- [RAD-0006] removed unused dependencies
- [RAD-0006] Fixed datadisplaypage double write
- [RAD-0006] Initial historical page done
- [RAD-0006] Initial template + lock orient
- Merge pull request #26 from Nntsyeo/RAD-0044
- [RAD-0044] Optimised google map reads
- Merge pull request #25 from Nntsyeo/RAD-0041
- [RAD-0041] Fix huge read on get last data
- add images
- Update Pod file
- Merge pull request #24 from Nntsyeo/RAD-0005
- [RAD-0005] changed to serverandcache
- [RAD-0005] Initial Google Map Ui
- [RAD-0005] Trying to incorporate sqflite
- [RAD-0005] Found out why circles not update
- [RAD-0005] Added initial color point and slider
- [RAD-0005] Initial added cloud data retrieval
- Merge pull request #23 from Nntsyeo/RAD-0038
- [RAD-0038] Added placeinfo and updated color
- fixed forced location service on login & infoplist
- Merge pull request #22 from Nntsyeo/ESP-0011
- Changed the battery indicator from float to int
- [ESP-0011] Fixed battery level detection
- modified by sean with jing zhi
- Merge commit 'e42166d08a87c1694bb922013eacf30edab08ce9' into ESP-0011
- added battery filtering feature
- updated eeprom and pwm-varying features
- added eeprom to store pwm.dutycycle
- modified for varying PWM duty cycle during operation
- added battery filtering feature
- Merge pull request #21 from Nntsyeo/RAD-0031
- [RAD-0031] Fixed bad element data creation issue
- [RAD-0031] Fixed cps not clip to 0 when disconnect
- [RAD-0031] Fixes user collection
- [RAD-0031] Adjusted database structure
- [RAD-0031] Fixed singlechild and other bugs
- Removed uSv/s and CPS, added battery voltage indicator and added feature to ignore double tap from user
- Added Cache and sub field in collection, ID is now consistent and continues from previous state
- UI modification and fixes for record button, scroll view and overflow fixes
- Fixed id reinstate when record state resume after pause
- id only increases when Record button in active state
- Restructured Cloud FireStore and id sequencing
- Merge pull request #19 from Nntsyeo/RAD-0023
- [RAD-0023] Fixed builds conflict & android bug
- [RAD-0023] Realtime graph, fixes visual bug
- [RAD-0023] Inital realtime graph
- [RAD-0023] Weird bluetooth bug
- [RAD-0023] Refactor code
- [RAD-0023] Initial graph plot visualizer
- Merge pull request #18 from Nntsyeo/RAD-0032
- [RAD-0032] Added bluetooth device unlink feature
- Merge pull request #17 from Nntsyeo/RAD-0021
- [RAD-0021] Finalized feature, fixed login bug
- Merge pull request #16 from Nntsyeo/RAD-0002
- Record only toggle every 60 seconds
- Readjust Timer and include Timeout timer
- Reset uSv accumalator
- Reset Acuumalting cpm
- Replaced Timer
- Fixed bugs for UI on Data Display Page and CPS parsing ? bug
- Added indicators and uSv/min and CPM windows
- [RAD-0002] Rebased this branch by fresh start
- Merge pull request #15 from Nntsyeo/ESP-0000
- modified for RGB status and duration
- Merge pull request #14 from Nntsyeo/RAD-0000
- [RAD-0000] Finalized changes on map and login
- [RAD-0000] Fixed login issue and revamp map
- [RAD-0000] Added Maps spp for IOS
- [RAD-0000] Added Gmap APIKEY
- [RAD-0000] Resolved conflict, Gmaps
- Merge pull request #13 from Nntsyeo/RAD-0004
- tested for deployment (success)
- Merge pull request #12 from Nntsyeo/RAD-0008
- Deleted the original icon for ios
- Merge pull request #11 from Nntsyeo/revert-10-RAD-0002
- Revert "Redirected X Code Build & Redesigned BLE pg"
- Merge pull request #10 from Nntsyeo/RAD-0002
- Redirected X Code Build & Redesigned BLE pg
- Merge pull request #8 from Nntsyeo/RAD-0001
- Added the name to the app
- Added Icon for the app
- [RAD-0001] Added general education page route
- Merge pull request #6 from Nntsyeo/RAD-0007
- Merge pull request #7 from Nntsyeo/RAD-0009
- [RAD-0007] Refactor code
- [RAD-0007] Fixed wrong align of icon in Android
- [RAD-0009] Fixed the login issue on android
- Merge pull request #5 from Nntsyeo/ESP-0000
- not sure why this happens...
- Added a TODO
- Merge pull request #4 from Nntsyeo/code_merge4
- Merged Education Page
- Merge pull request #2 from Nntsyeo/code_merge
- google map merge
- merge test 1
- Removed phone sign up method
- trying database file
- modified to ignore .DS_Store macOS file
- deleting sean_tryout_01 branch

mytask/lib/screens/login_screen.dart
import 'dart:async';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';
import 'package:mytask/config/config.dart';
import 'package:mytask/screens/email_pass_signup.dart';
import 'package:mytask/screens/home.dart';

class LoginScreen extends StatefulWidget {
  @override
  _LoginScreenState createState() => _LoginScreenState();
}

class _LoginScreenState extends State<LoginScreen> {

  final FirebaseAuth _auth = FirebaseAuth.instance;
  final Firestore _db = Firestore.instance;
  final TextEditingController _emailController = TextEditingController();
  bool _isLoading = false;
  final TextEditingController _passwordController = TextEditingController();
  bool _passwordVisible = false;
  final TextEditingController _resetController = TextEditingController();

  @override
  void dispose() {
    super.dispose();
    _passwordController.clear();
    _emailController.clear();
  }

  ///Reset Password
  @override
  Future<void> resetPassword(String email) async {
    await _auth.sendPasswordResetEmail(email: email);
  }

  Future<void> _signIn() async {
    String email = _emailController.text.trim();
    String password = _passwordController.text;
    String errorMessage;

    if (email.isNotEmpty && password.isNotEmpty) {
      try {
        FirebaseUser user = (await _auth.signInWithEmailAndPassword(
                email: email, password: password))
            .user;

        _db.collection("Users").document("Last Usage").setData({
          "email": email,
          "lastseen": DateTime.now(),
        });
        if (user.isEmailVerified) {
          setState(() {
            _isLoading = false;
          });
             Navigator.push(
                            context,
                            MaterialPageRoute(
                                builder: (context) => HomeScreen(
              email: user.email,
              name: user.displayName,
            )));
          showDialog(
              context: context,
              builder: (ctx) {
                return AlertDialog(
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(16),
                  ),
                  title: Text("Success"),
                  content: Text(
                      "Welcome to Rad - X"),
                  actions: <Widget>[
                    FlatButton(
                      child: Text("OK"),
                      onPressed: () {
                        _emailController.text = "";
                        _passwordController.text = "";
                        Navigator.of(ctx).pop();
                      },
                    ),
                  ],
                );
              });
        } else {
          setState(() {
            _isLoading = false;
          });
          showDialog(
              context: context,
              builder: (ctx) {
                return AlertDialog(
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(16),
                  ),
                  title: Text("Error"),
                  content: Text(
                      "Please verify email before signing in. Do you need a new verification link?"),
                  actions: <Widget>[
                    FlatButton(
                      child: Text("Cancel"),
                      onPressed: () {
                        Navigator.of(ctx).pop();
                      },
                    ),
                    FlatButton(
                      child: Text("OK"),
                      onPressed: () {
                        _emailController.text = "";
                        _passwordController.text = "";
                        user.sendEmailVerification();
                        Navigator.of(ctx).pop();
                      },
                    ),
                  ],
                );
              });
        }
      } catch (error) {
        switch (error.code) {
          case "ERROR_INVALID_EMAIL":
            errorMessage = "Your email address appears to be malformed.";
            break;
          case "ERROR_WRONG_PASSWORD":
            errorMessage = "Your password is wrong.";
            break;
          case "ERROR_USER_NOT_FOUND":
            errorMessage = "User with this email doesn't exist.";
            break;
          case "ERROR_USER_DISABLED":
            errorMessage = "User with this email has been disabled.";
            break;
          case "ERROR_TOO_MANY_REQUESTS":
            errorMessage = "Too many requests. Try again later.";
            break;
          case "ERROR_OPERATION_NOT_ALLOWED":
            errorMessage = "Signing in with Email and Password is not enabled.";
            break;
          default:
            errorMessage = "An undefined Error happened.";
        }
      }
      if (errorMessage != null) {
        setState(() {
          _isLoading = false;
        });
        showDialog(
            context: context,
            builder: (ctx) {
              return AlertDialog(
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(16),
                ),
                title: Text("Error"),
                content: Text("$errorMessage"),
                actions: <Widget>[
                  FlatButton(
                    child: Text("Cancel"),
                    onPressed: () {
                      Navigator.of(ctx).pop();
                    },
                  ),
                  FlatButton(
                    child: Text("OK"),
                    onPressed: () {
                      _emailController.text = "";
                      _passwordController.text = "";
                      Navigator.of(ctx).pop();
                    },
                  ),
                ],
              );
            });
      }
    } else {
      setState(() {
        _isLoading = false;
      });
      showDialog(
          context: context,
          builder: (ctx) {
            return AlertDialog(
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(16),
              ),
              title: Text("Error"),
              content: Text("Please provide email and password..."),
              actions: <Widget>[
                FlatButton(
                  child: Text("Cancel"),
                  onPressed: () {
                    Navigator.of(ctx).pop();
                  },
                ),
                FlatButton(
                  child: Text("OK"),
                  onPressed: () {
                    _emailController.text = "";
                    _passwordController.text = "";
                    Navigator.of(ctx).pop();
                  },
                ),
              ],
            );
          });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: _isLoading
          ? Center(
              child: CircularProgressIndicator(),
            )
          : SingleChildScrollView(
              child: Container(
                width: MediaQuery.of(context).size.width,
                height: 1.25 * MediaQuery.of(context).size.height,
                padding: EdgeInsets.only(
                  top: 40,
                ),
                child: Column(
                  children: <Widget>[
                    Container(
                      margin: EdgeInsets.only(
                        top: 60,
                      ),
                      child: Image(
                        image: AssetImage("assets/EIC2020_Logo.png"),
                        width: 200,
                        height: 150,
                      ),
                    ),
                    Container(
                      margin: EdgeInsets.only(
                        top: 10,
                      ),
                      child: Text(
                        "Login",
                        style: TextStyle(
                          fontSize: 30,
                        ),
                      ),
                    ),
                    Container(
                      padding: EdgeInsets.all(10),
                      margin: EdgeInsets.only(
                        top: 2,
                      ),
                      child: TextField(
                        controller: _emailController,
                        decoration: InputDecoration(
                          border: OutlineInputBorder(),
                          labelText: "Email",
                          hintText: "Write Email Here",
                        ),
                        keyboardType: TextInputType.emailAddress,
                      ),
                    ),
                    Container(
                      padding: EdgeInsets.all(10),
                      margin: EdgeInsets.only(
                        top: 2,
                      ),
                      child: TextField(
                        controller: _passwordController,
                        decoration: InputDecoration(
                            border: OutlineInputBorder(),
                            labelText: "Password",
                            hintText: "Write Password Here",
                            suffixIcon: IconButton(
                              onPressed: () {
                                setState(() {
                                  _passwordVisible = !_passwordVisible;
                                });
                              },
                              icon: _passwordVisible
                                  ? Icon(Icons.visibility)
                                  : Icon(Icons.visibility_off),
                            )),
                        obscureText: !_passwordVisible,
                      ),
                    ),
                    InkWell(
                      onTap: () {
                        setState(() {
                          _isLoading = true;
                        });
                        _signIn();
                      },
                      child: Container(
                        decoration: BoxDecoration(
                            gradient: LinearGradient(
                              colors: [primaryColor, secondaryColor],
                            ),
                            borderRadius: BorderRadius.circular(8)),
                        width: MediaQuery.of(context).size.width,
                        margin: EdgeInsets.symmetric(
                          horizontal: 30,
                          vertical: 20,
                        ),
                        padding: EdgeInsets.symmetric(
                          horizontal: 20,
                          vertical: 20,
                        ),
                        child: Center(
                            child: Text(
                          "Login",
                          style: TextStyle(
                            color: Colors.white,
                          ),
                        )),
                      ),
                    ),
                    FlatButton(
                      onPressed: () {
                        Navigator.push(
                            context,
                            MaterialPageRoute(
                                builder: (context) => EmailPassSignupScreen()));
                      },
                      child: Text("Signup using Email"),
                    ),
                    FlatButton(
                      onPressed: () {
                        showDialog(
                            context: context,
                            builder: (ctx) {
                              return AlertDialog(
                                shape: RoundedRectangleBorder(
                                  borderRadius: BorderRadius.circular(16),
                                ),
                                title: Text("Warning"),
                                content: Container(
                                  height: 200,
                                  child: Column(
                                    children: <Widget>[
                                      Text(
                                          "Are you sure you would like to reset your password?"),
                                      SizedBox(height: 10),
                                      TextField(
                                        controller: _resetController,
                                        decoration: InputDecoration(
                                          border: OutlineInputBorder(),
                                          labelText: "Email",
                                          hintText: "Write Email Here",
                                        ),
                                        keyboardType:
                                            TextInputType.emailAddress,
                                      ),
                                      SizedBox(height: 10),
                                      Row(
                                        mainAxisAlignment:
                                            MainAxisAlignment.center,
                                        children: <Widget>[
                                          FlatButton(
                                            child: Text("Cancel"),
                                            onPressed: () {
                                              Navigator.of(ctx).pop();
                                            },
                                            shape: RoundedRectangleBorder(
                                                side: BorderSide(
                                                    color: Colors.blueGrey,
                                                    width: 2,
                                                    style: BorderStyle.solid),
                                                borderRadius:
                                                    BorderRadius.circular(30)),
                                          ),
                                          SizedBox(width: 40),
                                          FlatButton(
                                            child: Text("Yes"),
                                            onPressed: () {
                                              String reset =
                                                  _resetController.text;
                                              if (reset.isNotEmpty)
                                                resetPassword(
                                                    _resetController.text);
                                              showDialog(
                                                  context: context,
                                                  builder: (ctx) {
                                                    return AlertDialog(
                                                      shape:
                                                          RoundedRectangleBorder(
                                                        borderRadius:
                                                            BorderRadius
                                                                .circular(16),
                                                      ),
                                                      title: Text(
                                                          "Reset Password"),
                                                      content: Text(
                                                          "An email has been sent to your account to reset your password"),
                                                      actions: <Widget>[
                                                        FlatButton(
                                                          child: Text("OK"),
                                                          onPressed: () {
                                                            Navigator.of(ctx)
                                                                .pop();
                                                          },
                                                        ),
                                                      ],
                                                    );
                                                  });
                                            },
                                            shape: RoundedRectangleBorder(
                                                side: BorderSide(
                                                    color: Colors.blueGrey,
                                                    width: 2,
                                                    style: BorderStyle.solid),
                                                borderRadius:
                                                    BorderRadius.circular(30)),
                                          ),
                                        ],
                                      ),
                                    ],
                                  ),
                                ),
                              );
                            });
                      },
                      child: Text("Reset Password"),
                    ),
                  ],
                ),
              ),
            ),
    );
  }
}
